src/util/kaldi-table-inl.h

```diff
@@ -178,25 +191,40 @@ template<class Holder>  class SequentialTableReaderScriptImpl:
       else  // orig_state_ was kLoadFailed, which only could have happened
         // if the user called FreeCurrent().
         KALDI_ERR << "You called Value() after FreeCurrent().";
-    } else if (state_ != kLoadSucceeded) {
+    } else if (state_ == kRangeExtractionFailed) {
+      KALDI_ERR << "Failed to load object from "
+                << PrintableRxfilename(data_rxfilename_) 
+                << " [" << range_ << "] "
+                << " (to suppress this error, add the permissive "
+                << "(p, ) option to the rspecifier.";
+    } else if (state_ != kLoadSucceeded && state_ != kRangeExtracted) {
       // This would be a coding error.
       KALDI_ERR << "Value() called at the wrong time.";
     }
-    return holder_.Value();
+    if (state_ == kRangeExtracted) {
+      return object_;
+    } else {
+      // Here state_ is kLoadSucceeded.
+      return holder_.Value();
+    }
   }
+
   void FreeCurrent() {
-    if (state_ == kLoadSucceeded) {
+    if (state_ == kLoadSucceeded || 
+        state_ == kRangeExtracted || state_ == kRangeExtractionFailed) {
       holder_.Clear();
       state_ = kLoadFailed;
     } else {
       KALDI_WARN << "FreeCurrent called at the wrong time.";
     }
   }
+
   void SwapHolder(Holder *other_holder) {
     // call Value() to ensure we have a value, and ignore its return value while
     // suppressing compiler warnings by casting to void.
     (void) Value();
-    if (state_ == kLoadSucceeded) {
+    if (state_ == kLoadSucceeded ||
+        state_ == kRangeExtracted || state_ == kRangeExtractionFailed) {
       holder_.Swap(other_holder);
```

Comment:
ok thanks

On Sun, Mar 27, 2016 at 12:55 AM, Daniel Povey <notifications@github.com>
wrote:

> In src/util/kaldi-table-inl.h
> <https://github.com/kaldi-asr/kaldi/pull/611#discussion_r57523436>:
>
> >    void SwapHolder(Holder *other_holder) {
> >      // call Value() to ensure we have a value, and ignore its return value while
> >      // suppressing compiler warnings by casting to void.
> >      (void) Value();
> > -    if (state_ == kLoadSucceeded) {
> > +    if (state_ == kLoadSucceeded ||
> > +        state_ == kRangeExtracted || state_ == kRangeExtractionFailed) {
> >        holder_.Swap(other_holder);
>
> "bg" is where for sequential readers, if you add ",bg" to the rspecifier
> it does the reading in a background thread (useful in processes that use
> GPUs, to get fuller usage).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/611/files/d778589ed02590c92d5578281dd181f22280fa8b#r57523436>
>



-- 
Xiao-hui